pkg/tls/certificatemanagement/controller/ca.go

```diff
@@ -0,0 +1,533 @@
+	if cm.CertificateManagement() == nil && len(certPem) != 0 {
+		x509Cert, err := parseCertificate(certPem)
+		if err != nil {
+			return false
+		}
+		return string(x509Cert.AuthorityKeyId) == string(cm.Certificate.AuthorityKeyId)
+	}
+	return false
+}
+
+// GetCertificate returns a Certificate. If the certificate is not found a k8s.io NotFound error is returned.
+func (cm *certificateManager) GetCertificate(cli client.Client, secretName, secretNamespace string) (render.Certificate, error) {
+	if cm.CertificateManagement() != nil {
+		return NewCertificate(secretName, nil), nil
+	}
+	secret := &corev1.Secret{}
+	err := cli.Get(context.Background(), types.NamespacedName{
+		Name:      secretName,
+		Namespace: secretNamespace,
+	}, secret)
+	if err != nil {
+		return nil, err
+	}
+	_, cert := getKeyCertPEM(secret)
+	if len(cert) == 0 {
+		return nil, fmt.Errorf("secret %s/%s is missing a certificate", secret.Namespace, secret.Name)
+	}
+	x509Cert, err := parseCertificate(cert)
+	if err != nil {
+		return nil, err
+	}
+	if x509Cert.NotAfter.Before(time.Now()) || x509Cert.NotBefore.After(time.Now()) {
+		return nil, fmt.Errorf("secret %s is not valid at this date", secretName)
+	}
+	return NewCertificate(secretName, cert), nil
+}
+
+// GetKeyPair returns an existing KeyPair. If the KeyPair is not found, a k8s.io NotFound error is returned.
+func (cm *certificateManager) GetKeyPair(cli client.Client, secretName, secretNamespace string) (render.KeyPair, error) {
+	secret := &corev1.Secret{}
+	err := cli.Get(context.Background(), types.NamespacedName{
+		Name:      secretName,
+		Namespace: secretNamespace,
+	}, secret)
+	if err != nil {
+		if !kerrors.IsNotFound(err) && cm.keyPair.CertificateManagement != nil {
+			return certificateManagementKeyPair(cm, secretName, nil), nil
+		}
+		return nil, nil
+	}
+	keyPEM, certPEM := getKeyCertPEM(secret)
+	if len(keyPEM) == 0 || len(certPEM) == 0 {
+		return nil, fmt.Errorf("secret %s/%s is missing a certificate", secret.Namespace, secret.Name)
+	}
+	cert, err := parseCertificate(secret.Data[corev1.TLSCertKey])
+	if err != nil {
+		return nil, err
+	}
+	if cert.NotAfter.Before(time.Now()) || cert.NotBefore.After(time.Now()) {
+		return nil, fmt.Errorf("secret %s is not valid at this date", secretName)
+	}
+	return &keyPair{
+		ca:             cm,
+		name:           secretName,
+		privateKeyPem:  keyPEM,
+		certificatePEM: certPEM,
+	}, nil
+}
+
+// CertificateManagement returns the CertificateManagement object or nil if it is not configured.
+func (cm *certificateManager) CertificateManagement() *operatorv1.CertificateManagement {
+	return cm.keyPair.CertificateManagement
+}
+
+type keyPair struct {
+	csrImage       string
+	name           string
+	privateKeyPem  []byte
+	certificatePEM []byte
+	clusterDomain  string
+	*operatorv1.CertificateManagement
+	dnsNames                 []string
+	useCertificateManagement bool
+	ca                       render.CertificateManager
```

Comment:
This is highly suspect having the mutual depedendency between keyPai and the certificate manager... that very tight coupling.

Reply:
This seems to only be used for the check if this is BYO, so why not just set that in the constructor.